### PR TITLE
explain aggressive limit

### PR DESCRIPTION
**Pull request by SergeyKanzhelev**
#109

### CODE REVIEW

File: trace_context/HTTP_HEADER_FORMAT.md [hunk 1]
```diff
@@ -155,7 +155,9 @@ Rather, the entry would be rewritten to only include the most recent position:
 `congo=congosSecondPosition,rojo=rojosFirstPosition`
 
 **Limits:**
-Maximum length of a combined header MUST be less than 512 bytes. 
+Maximum length of a combined header MUST be less than 512 bytes.
+
+`tracestate` field contains essential information for requests correlation. Platforms and tracing systems MUST propagate this header. Compliance with specification will require storing of `tracestate` as part of request payload or associated metadata. Allowing the long field values can make compliance to the specification impossible. Thus the aggressive limit of 512 bytes was chosen.
```
Comment (codefromthecrypt) [inline trace_context/HTTP_HEADER_FORMAT.md]: I still think a more concrete example of why 512 vs 2048 or any other fixed val

Comment (codefromthecrypt) [reply trace_context/HTTP_HEADER_FORMAT.md]: @codingfabian @abhiksingh can you help explain this? I think you have a lot of experience with systems, maybe can help shed some light. Right now, feels just like a couple people saying things again and again.. we need better than this.

In http most servers are subject to a total limit of header size, not individual. Other parts of the spec later will overrun completely towards the same budget. We can assume that this size is indeed only about this part and not directly about headers in general.

I think guidance should clearly be for vendors to use the least space practical. However, where we put a hard limit, such a line is important, as it affects the viability of multiple vendors at all. IOTW, it shouldn't be cripplingly small, neither should it arbitrarily and early punt towards continuations (like a second header with the same name) as that just moves the problem in ways that only work in systems with multi-value headers.

So anyway, any concrete examples to help pin down what I still feel is a bit arbitrary?

Comment (SergeyKanzhelev) [reply trace_context/HTTP_HEADER_FORMAT.md]: The history of 512 was that it was a number nobody wanted to go below. I heard another suggestion is to haver total of `tracestate` and `traceparent` be under 1K.

CC: @bogdandrutu who has opinion and example with some queue service and metadata.

Comment (AloisReitbauer) [reply trace_context/HTTP_HEADER_FORMAT.md]: I would rephrase this part to be a bit more open. I suggest we define this as:

Tracestate must be forwared by all participating tracing if it i is less or equal than 512 bytes. If the tracestate is bigger than this, the tracer can decide to forward the trace state or drop it.

Note: I know we talked about truncation. However, since there is no final decision on how this would work, the current working draft should state that you either forward or drop.

Comment (CodingFabian) [reply trace_context/HTTP_HEADER_FORMAT.md]: 👍 on the wording by @AloisReitbauer

**Comment by tedsuo** [comment]
I believe the goal of 512k was to ensure that 3 hops (the maximum expected in most real world scenarios) would not cause this header to be dropped or truncated. If that is the motivation, it should be listed here. Currently this explains that an overflow could occur, but not why the limit is 512k vs something else.

If such an explanation is better put in the Rationale document instead, ignore this.

**Review by mtwo — APPROVED** [review]

**Comment by codefromthecrypt** [comment]
512 bytes you meant :P

agree it is still arbitrary and would like to see concretely why not 1024 or something larger. This is the very best time to answer this question rather than people years later looking for the answer and never finding it. If we can't answer the question we shouldn't have this constraint.

**Comment by jkowall** [comment]
I know we will overrun this size... I'm not sure what the number is but 1024 seems small to me. 512 is a big issue.

**Comment by yurishkuro** [comment]
This is easily half a dozen times I've seen this question asked: why 512? Where did it come from? Did someone just randomly picked it?

**Comment by SergeyKanzhelev** [comment]
@adriancole even if there will be no meaningful explanation we need to pick a limit. One of the purpose of this limit is to make sure that `tracestate` is not abused for a baggage. It is critical for services to propagate it and ideally store with produced telemetry without implication of cost or reliability attacks. So it's better if nobody assumes "no limit" from the first iteration. People years later will think we did it wrong anyway =).

(not advocating for complete random here so let's discuss if there are implications of `512` to be a limit).

@jkowall what data do you plan to store in `tracestate` to overrun the limit? Note, `tracestate` is intended for information crucial for positioning of the span in a distributed call graph. Baggage is propagated via separate header.

@yurishkuro - yes, it was mostly random pick. Mostly it was coming from discussion that `tracestate` when implemented by cloud services - should be stored and propagated thru metadata. Ideally free of charge and not affecting performance.

**Comment by codefromthecrypt** [comment]
"People years later will think we did it wrong anyway =)." lol true

**Comment by SergeyKanzhelev** [comment]
BTW, I realize this PR doen't close the corresponding issue. It merely explains why value needs to have an aggressive limit. Not the reasoning for `512`.

**Comment by jkowall** [comment]
@SergeyKanzhelev I'm just saying out header can get quite large, and may overrun this number. The data in the header is crucial for us to place the span in context of the application, business transaction, and other data we require. Some of that could be moved to other headers but then it may not be propagated making it useless for downstream correlation.

**Comment by wu-sheng** [comment]
> @SergeyKanzhelev I'm just saying out header can get quite large, and may overrun this number. The data in the header is crucial for us to place the span in context of the application, business transaction, and other data we require. Some of that could be moved to other headers but then it may not be propagated making it useless for downstream correlation.

Saw and known some header of AppD, those must be useful for backend analysis. SkyWalking depends on those headers too. Just not so long.

I think @SergeyKanzhelev hope to move those headers to the third type of header. But as same as SkyWalking, I guess the backends need those long headers. Those are not optional. :)

**Comment by SergeyKanzhelev** [comment]
@wu-sheng I understand you are talking about hierarchical identifier, correct? Do you have an acceptable size limit in mind to satisfy the requirement of keeping `tracestate` small to guarantee propagation by cloud vendors?

**Comment by wu-sheng** [comment]
>  I understand you are talking about hierarchical identifier, correct? 

In AppD, more like hierarchical path, or trace path. At least last time I read the headers.  Of course, @jkowall is the authority :)

I have no good suggestions... sorry. I am just aware a lot of commercial APMs put a lot of info in headers.

**Comment by codefromthecrypt** [comment]
here's a point of reference. Amazon will restart its trace ID if the header
gets to 7KiB

https://docs.aws.amazon.com/elasticloadbalancing/latest/application/load-balancer-request-tracing.html
